src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java

```diff
@@ -208,4 +252,112 @@ else if (!testMessages.isEmpty()) {
+        Set<ProgrammingExerciseTestCase> testCasesFromFeedbacks = getTestCasesFromBuildResult(buildResult, exercise);
+        // Get test cases that are not already in database - those will be added as new entries.
+        Set<ProgrammingExerciseTestCase> newTestCases = testCasesFromFeedbacks.stream().filter(testCase -> existingTestCases.stream().noneMatch(testCase::isSameTestCase))
+                .collect(Collectors.toSet());
+        // Get test cases where the activate state has changed (re-added or removed tests).
+        Set<ProgrammingExerciseTestCase> testCasesWithUpdatedActivation = getTestCasesWithUpdatedActivation(existingTestCases, testCasesFromFeedbacks);
+
+        Set<ProgrammingExerciseTestCase> testCasesToSave = new HashSet<>();
+        testCasesToSave.addAll(newTestCases);
+        testCasesToSave.addAll(testCasesWithUpdatedActivation);
+
+        setTestCaseType(testCasesToSave, exercise.getProgrammingLanguage());
+
+        // Ensure no duplicate TestCase is present: TestCases have to have a unique name per exercise.
+        // Just using the uniqueness property of the set is not enough, as the equals/hash functions
+        // compares test cases by their id.
+        testCasesToSave.removeIf(candidate -> testCasesToSave.stream().filter(testCase -> testCase.getTestName().equalsIgnoreCase(candidate.getTestName())).count() > 1);
+
+        if (!testCasesToSave.isEmpty()) {
+            testCaseRepository.saveAll(testCasesToSave);
+            programmingExerciseTaskService.updateTasksFromProblemStatement(exercise);
+            // Update the test cases in the problem statement.
+            // This handles the case if the problem statement already contains the name of test case
+            // that got later pushed into the test repository. Since this test case now exists,
+            // the problem statement should now refer to its id.
+            programmingExerciseTaskService.replaceTestNamesWithIds(exercise);
+            programmingExerciseRepository.save(exercise);
+            return true;
+        }
+        return false;
+    }
+
+    private Set<ProgrammingExerciseTestCase> getTestCasesWithUpdatedActivation(Set<ProgrammingExerciseTestCase> existingTestCases,
+            Set<ProgrammingExerciseTestCase> testCasesFromFeedbacks) {
+        // We compare the new generated test cases from feedback with the existing test cases from the database
+        return existingTestCases.stream().filter(existing -> {
+            Optional<ProgrammingExerciseTestCase> matchingTestCase = testCasesFromFeedbacks.stream().filter(existing::isSameTestCase).findFirst();
+            // Either the test case was active and is not part of the feedback anymore OR was not active before and is now part of the feedback again.
+            return (matchingTestCase.isEmpty() && existing.isActive()) || (matchingTestCase.isPresent() && matchingTestCase.get().isActive() && !existing.isActive());
```

Comment:
That case needs to be updated and also gets updated. Either a test is no longer present (active -> non active) or was inactive and is now again present in the build result.

Reply:
So `matchingTestCase.isPresent() && !matchingTestCase.get().isActive()` will always resolve to false?

Reply:
oh, I see where the confusion comes from. `matchingTestCase.get().isActive()` does not need to be checked, this will always be true when the test is part of the solution feedback.
![grafik](https://github.com/ls1intum/Artemis/assets/26540346/378c0391-2cbf-4739-8077-09bfe4e6f62c)